lib/hyacinth/digital_object/type_def/dynamic_field_data.rb

```diff
@@ -0,0 +1,117 @@
+# frozen_string_literal: true
+
+module Hyacinth
+  module DigitalObject
+    module TypeDef
+      class DynamicFieldData < Hyacinth::DigitalObject::TypeDef::JsonSerializableCollection
+        TYPE_TO_FORM_TYPE = {
+          rights_metadata: ['item_rights', 'asset_rights'],
+          descriptive_metadata: ['descriptive']
+        }.freeze
+
+        def initialize(type)
+          super()
+          raise ArgumentError, "type '#{type}' is invalid" unless TYPE_TO_FORM_TYPE.keys.map(&:to_s).include?(type.to_s)
+
+          @type = type.to_sym
+        end
+
+        def from_serialized_form_impl(json_var)
+          return nil if json_var.nil?
+          raise ArgumentError, "Expected hash, but got: #{json_var.class}" unless json_var.is_a?(Hash)
+
+          terms_map = extract_terms(field_map, json_var)
+
+          search_query = terms_map.map { |vocab, terms|
+            uris = terms.map { |t| "\"#{t['uri']}\"" }.compact.join(' OR ')
+            "(vocabulary:\"#{vocab}\" AND uri:(#{uris}))"
+          }.join(' OR ')
+
+          search_results = Hyacinth::Config.term_search_adapter.search(post: true) do |solr_params|
+            solr_params.q(search_query, escape: false)
+          end
+
+          # Retrieve all custom fields for vocabularies so we know which fields should be add to the hash.
+          # This ensure that there's always a value for the all the custom fields defined and does not add
+          # any custom fields that have been removed from the Vocabulary since indexing.
+          vocab_to_custom_fields = Vocabulary.where(string_key: terms_map.keys)
+                                             .map { |v| [v.string_key, v.custom_fields] }
+                                             .to_h
+
+          terms_map.each do |vocab, terms|
+            custom_field_definitions = vocab_to_custom_fields[vocab]
+
+            terms.each do |term_hash|
+              solr_term_hash = search_results['response']['docs'].find do |result|
+                result['uri'] == term_hash['uri'] && result['vocabulary'] == vocab
+              end
+
+              next if solr_term_hash.blank?
+
+              full_term = {}
+
+              { 'authority' => '', 'pref_label' => '', 'alt_labels' => [], 'term_type' => '' }.map do |key, default|
```

Comment:
Do we actually want a term_type default of `''`?  I think if a term comes in with no type, that would cause all sorts of problems and an empty string wouldn't make it any better.  Maybe not necessary to worry about that default here.

Reply:
I realize that some of these fields will never be empty, but I am just trying to assure that the empty fields do have a value. What's your suggestion? Should I just pull the fields that are always populated out of the logic and merge them separately?

Reply:
I was just suggesting that we remove `'term_type' => ''` from this hash since I didn't think it would do anything.

Reply:
It is fetching the key and moving it over to the full_term hash, so I can find an alternate way to do that for the fields that should always be populated.

Reply:
Sounds good!